include/small/util.h

```diff
@@ -58,3 +113,271 @@ small_getpagesize(void)
 		return 4096;
 	return page_size;
 }
+
+/**
+ * Align a size - round up to nearest divisible by the given alignment.
+ * Alignment must be a power of 2
+ */
+static inline size_t
+small_align(size_t size, size_t alignment)
+{
+	/* Must be a power of two */
+	assert((alignment & (alignment - 1)) == 0);
+	/* Bit arithmetics won't work for a large size */
+	assert(size <= SIZE_MAX - alignment);
+	return (size - 1 + alignment) & ~(alignment - 1);
+}
+
+/**
+ * Align value to the nearest divisible by the given alignment which is
+ * not greater than value. Alignment must be a power of 2.
+ */
+static inline size_t
+small_align_down(size_t value, size_t alignment)
+{
+	assert((alignment & (alignment - 1)) == 0);
+	return value & ~(alignment - 1);
+}
+
+/** Round up a number to the nearest power of two. */
+static inline size_t
+small_round(size_t size)
+{
+	if (size < 2)
+		return size;
+	assert(size <= SIZE_MAX / 2 + 1);
+	assert(size - 1 <= ULONG_MAX);
+	size_t r = 1;
+	return r << (sizeof(unsigned long) * CHAR_BIT -
+		     __builtin_clzl((unsigned long) (size - 1)));
+}
+
+/** Binary logarithm of a size. */
+static inline size_t
+small_lb(size_t size)
+{
+	assert(size <= ULONG_MAX);
+	return sizeof(unsigned long) * CHAR_BIT -
+		__builtin_clzl((unsigned long) size) - 1;
+}
+
+#ifdef SMALL_MALLOC_IMPL
+
+/**
+ * Small wrapper allows to hold extra information (header) for every user
+ * memory allocation. For this purpose we allocate more memory then requested
+ * and keep a header and user data in a same chunk of memory. So wrapper memory
+ * layout is next:
+ *
+ *     UUU HHHHHHH AAA PPPPPPPP UUUU
+ *
+ * Where
+ *    P - user payload
+ *    H - header
+ *    A - memory used for alignment
+ *    F - unused memory (raised to header and payload alignments)
+ *
+ * Header has SMALL_HEADER_ALIGNMENT (large alignment to allow any
+ * types in header). User payload is aligned to given alignment A.
+ * Additionally payload is not aligned on 2 * A. This improves unaligned
+ * access check. Header should be inherited from struct small_header.
+ *
+ * Wrapper is used with ASAN builds. Thus to check buffer underflows/overflows
+ * wrapper memory is poisoned except for user payload of course. Note that
+ * due to poison alignment restrictions we may not be able to poison
+ * part of alignment area before payload. In this case we write magic to
+ * the area which cannot be poisoned and check it is not changed when memory
+ * is freed.
+ */
+
+enum {
+	/* Alignment of header in wrapper. */
+	SMALL_HEADER_ALIGNMENT = sizeof(long long),
+	/*
+	 * ASAN does not allow to precisely poison arbitrary ranges of memory.
+	 * However if range end is 8-aligned or range end is end of memory
+	 * allocated with malloc() then poison is precise.
+	 */
+	SMALL_POISON_ALIGNMENT = 8,
+};
+
+/** Random magic to be used for memory that cannot be poisoned. */
+extern uint64_t small_magic;
```

Comment:
Nit: It doesn't need to be an extern variable. It could be a macro constant.

Reply:
Done.